include/SDL3/SDL_gpu.h

```diff
@@ -0,0 +1,797 @@
+ * SDL. These are not all guaranteed to _work_ at runtime, but they are the
+ * backends that are potentially available. You do not have to iterate
+ * these; if you request a NULL driver to SDL_GpuCreateDevice, SDL will try
+ * to make the best decision for the current system, but if you want to
+ * try multiple drivers or log them for debugging, etc, here you go.
+ *
+ * The list of drivers is static and based on what was compiled into SDL; it
+ * does not change between calls to these functions.
+ */
+int SDL_GetNumGpuDrivers(void);
+const char *SDL_GetGpuDriver(int index);
+
+/* !!! FIXME: Enumerate physical devices. Right now this API doesn't allow it. */
+
+typedef struct SDL_GpuDevice SDL_GpuDevice;
+SDL_GpuDevice *SDL_CreateGpuDevice(const char *label, const char *driver);  /* `label` is for debugging, not a specific device name to access. */
+void SDL_DestroyGpuDevice(SDL_GpuDevice *device);
+
+/* !!! FIXME: device caps */
+
+/* CPU buffers live in RAM and can be accessed by the CPU. */
+typedef struct SDL_CpuBuffer SDL_CpuBuffer;
+SDL_CpuBuffer *SDL_CreateCpuBuffer(const char *label, SDL_GpuDevice *device, const Uint32 buflen, const void *data);
+void SDL_DestroyCpuBuffer(SDL_CpuBuffer *buffer);
+void *SDL_LockCpuBuffer(SDL_CpuBuffer *buffer, Uint32 *_buflen);
+int SDL_UnlockCpuBuffer(SDL_CpuBuffer *buffer);
+
+/*
+ * GPU buffers live in GPU-specific memory and can not be accessed by the CPU.
+ *  If you need to get data to/from a GPU buffer, encode a blit operation
+ *  to move it from/to a CPU buffer. Once in a CPU buffer, you can lock it to access data in your code.
+ * There is no initial `data` pointer here, like CPU buffers have, since there's significantly more
+ *  work to initialize them. If you plan to do an upload to the GPU buffer at the same time as
+ *  you create it, there's a convenience function that will do it at the cost of blocking until the
+ *  upload is complete: SDL_GpuCreateAndInitBuffer
+ */
+typedef struct SDL_GpuBuffer SDL_GpuBuffer;
+SDL_GpuBuffer *SDL_CreateGpuBuffer(const char *label, SDL_GpuDevice *device, const Uint32 buflen);
+void SDL_DestroyGpuBuffer(SDL_GpuBuffer *buffer);
+
+
+typedef enum SDL_GpuTextureType
+{
+    SDL_GPUTEXTYPE_1D,
+    SDL_GPUTEXTYPE_2D,
+    SDL_GPUTEXTYPE_CUBE,
+    SDL_GPUTEXTYPE_3D,
+    SDL_GPUTEXTYPE_1D_ARRAY,
+    SDL_GPUTEXTYPE_2D_ARRAY,
+    SDL_GPUTEXTYPE_CUBE_ARRAY
+} SDL_GpuTextureType;
+
+typedef enum SDL_GpuPixelFormat
+{
+    SDL_GPUPIXELFMT_INVALID,
+    SDL_GPUPIXELFMT_B5G6R5,
+    SDL_GPUPIXELFMT_BGR5A1,
+    SDL_GPUPIXELFMT_RGBA8,
+    SDL_GPUPIXELFMT_RGBA8_sRGB,
+    SDL_GPUPIXELFMT_BGRA8,
+    SDL_GPUPIXELFMT_BGRA8_sRGB,
+    SDL_GPUPIXELFMT_Depth24_Stencil8
+    /* !!! FIXME: some sort of YUV format to let movies stream efficiently? */
+    /* !!! FIXME: s3tc? pvrtc? other compressed formats? We'll need a query for what's supported, and/or guarantee it with a software fallback...? */
```

Comment:
I was hoping to avoid this, but there isn't a lot of agreement about what compressed texture types are universally available, so yeah, I think we'll be adding this.

Reply:
Unfortunately there's no such thing as "universally available" when it comes to compressed textures. On the desktop that would probably be S3TC, and probably BPTC/BC7 at the feature level you're targeting. But those are rarely, if ever, supported by mobile GPUs. Those often have ETC/ETC2 (OpenGL ES 3 requires those) and probably ASTC. ETC2 support is also required for desktop GL 4.3 as part of [ARB_ES3_compatibility](https://registry.khronos.org/OpenGL/extensions/ARB/ARB_ES3_compatibility.txt), but that's usually implemented as a software fallback. This sad state of affairs is why Basis Universal exists in the first place.